### ISSUE TITLE
Lets use Tarn.js as a pool instead of generic-pool

### ISSUE THREAD

**Issue by elhigu** [tgriesser/knex#2339]
# Environment

Knex version: > 0.8
Database + version: all
OS: all

# Feature discussion / request

Knex uses currently `generic-pool`, which has been source of multiple problems and it is not very actively maintained (author is not replying to issues / pull requests). I opened this issue, after I noticed that the `testOnBorrow` feature, which is needed by `knex` didn't have any tests (had one that was just commented out).

Also the `generic-pool` is lacking tests, which means we cannot trust that it works correctly when it is updated. For example `testOnBorrow` feature which should be used to prevent returning dead connections (e.g. closed by database)

I would like to open up discussion https://github.com/tgriesser/knex/pull/1702 again about using own pool component  [Tarn.js](https://github.com/Vincit/tarn.js), which we can update / fix when ever there is need and whose API can be extended to support `knex` usage so that we don't need to write work arounds to support `generic-pool`.

Some related issues: 
https://github.com/tgriesser/knex/issues/2321
https://github.com/tgriesser/knex/issues/1833
https://github.com/tgriesser/knex/issues/2325

**Comment by wubzz** [comment]
From briefly looking at `Tarn.js` docs it seems to cover everything we're currently relying on from `generic-pool`, even _(almost)_ under the same config syntax, which is great. I have to agree that every upgrade of `generic-pool` has resulted in at least one or more major bugs in a release, and for that reason I strongly support this request.

I think for me `generic-pool` crossed the line when the [upgrade guide](https://gist.github.com/sandfox/5ca20648b60a0cb959638c0cd6fcd02d) from version `2` -> `3` failed to mention the complete removal of default `reapIntervalMillis` which heavily breaks existing code... This in turn resulted in a `knex` docs change/warning.

@tgriesser You were against this previously in hopes of instead having the user supply method of pooling. While that is still a good idea, perhaps we should still make one final "swap" of pool libs before that time comes?

**Comment by richraid21** [comment]
If this change is going to occur, it might be a great moment to add support for different read/write endpoints. For high availability systems, this need is crucial.  

Accomplishing this at the pool level would be the most robust solution.

**Comment by elhigu** [comment]
@richraid21 that is a separate issue, better to talk about it in its own ticket https://github.com/tgriesser/knex/issues/2253. Starting do specify there how that master/slave configuration should actually work and what kind of APIs knex would provide for it would be a good start.

**Comment by tuananh** [comment]
have an upvote. however, when i look at tarn.js, it's even less active. last update was 1 year ago

**Comment by elhigu** [comment]
@tuananh Its not active, because it was created for knex to be ~drop in replacement for generic-pool. If it is changed, then it will be maintained by the same maintainers with knex and objection.js.

**Comment by tuananh** [comment]
@elhigu can you get in touch with them first? what if they don't want to commit to that project now.

**Comment by elhigu** [comment]
@tuananh I work at the same company (actually in the same room) @Vincit with @koskimas who wrote that pool replacement module for knex (who is also author of objection.js ORM).

I’ll add full access to Tarn for me and for the other knex collaborators. Our employee also supports this kind of open source work so we can put up to 20 hours per month to it.

**Comment by koskimas** [comment]
@elhigu You mean our employ**er**, not employ**ee** 😄

**Comment by pbarbiero** [comment]
Generic-pool has been causing dramatic issues for us in production as well, randomly and without reason. Things like days go by, then we have about a 40% failure rate in acquiring connections from the pool which forces us to reboot the application (which is hard to do to a production service). We would roll back to knex 0.13, but we need mssql v4.1 which adds support for multiSubnetFailover.

Any chance this can be looked into? knex 0.14 is VERY unstable due to generic-pool v3. I would be happy to share some impact reports and the amount of headache we are going through if it would help push this request over the edge.

**Comment by wubzz** [comment]
In hindsight I suppose we could have reverted #2208 and published a new version long ago given how unstable 0.14 is at its current state.

What do you think @elhigu @tgriesser ?

**Comment by elhigu** [comment]
@wubzz I'm working to fix this pool thing with  with tarn author (@koskimas) once and for all. Starting by writing more tests for current mysql and memory consumption problems and reviewing all the knex pool code + dialects. 

I'd rather not go back to older generic-pool which also had its problems and reverting back needs also some effort.

**Comment by wubzz** [comment]
@elhigu That sounds like a good plan and I'm all for it. However, I also expect it will be quite time consuming since it has to be done with caution. For that reason it might still be valid to revert #2208 so that the default `npm i knex` doesn't break applications the way that it currently is.

Reverting the PR should not require much effort (afaik) since all changes were essentially "internal" knex changes, not to the actual pool options (aside from `evictionRunIntervalMillis` which was new). Perhaps @jstewmon can clarify on that.

Also may have to take into consideration @tgriesser [comment](https://github.com/tgriesser/knex/issues/2383#issuecomment-352461273) relating to defaulting to dialect built-in pooling so that the option is still on the table, so to speak.

**Comment by elhigu** [comment]
@wubzz reverting back needs to be done very carefully and there were lots of changes and there are some more tests now which may fail with old pool. Original upgrade to new pool version was not done very carefully I spent tens of hours debuggin / fixing it afterwards.

If we are going to use dialect builtin pools there probably needs to be some adapter layer anyways for knex to be able to interract with all the different ones and for drivers that does not have their own pool some default pool is needed anyways (IIRC even sqlite supports multiple connections nowadays). So I see that as a separate issue. 

Also if dialect specific pools support is implemented I'm expecting lots of weird problems from the fact that they work in different ways. One more problem with it is that we have no idea how we would be supporting those dialects builtin pools (probably it needs at least big changes to configuration) and I don't know anyone who would be willing to implement it.

To summarise: I don't see dialect builtin pooling support being done anytime soon and downgrade to older generic-pool is not as easy as it sounds and it might not even fix that many problems. We are able to update to Tarn and fix all the issues in Tarn a lot faster. If someone sends PR that downgrades knex to use older generic-pool we can try it out though.

**Comment by wubzz** [comment]
I don't completely understand how or why new features and/or tests would be dependant on generic-pool v3 instead of v2, but I guess I might not see the whole picture.

I would argue that using the default dialect implementation of pooling is always the way to go. It is well tested and optimized for that specific dialect, what are the downsides? For dialects that don't have their own pooling system we could default to an external lib (such as Tarn.js) as per @tgriesser previous comment.

In general I believe knex puts down too much work towards handling pooling. We have swapped pooling libs several times already. Why not rely on something that is already well tested?

@tgriesser Can you weigh in on what you think the future should hold for pooling connections in knex?

**Comment by koskimas** [comment]
@pbarbiero I created #2450 if you'd like to test if that fixes your pooling problems.

**Comment by elhigu** [comment]
> I don't completely understand how or why new features and/or tests would be dependant on generic-pool v3 instead of v2, but I guess I might not see the whole picture.

It is completely possible that things work well and easily. I just might be too pessimistic about it.

> I would argue that using the default dialect implementation of pooling is always the way to go. It is well tested and optimized for that specific dialect, what are the downsides? For dialects that don't have their own pooling system we could default to an external lib (such as Tarn.js) as per @tgriesser previous comment.

I see that too as a good option eventually, but it needs planning and work to get there. I don't know how much since I haven't looked how it could possibly work (AFAIK there are no one who has).

> In general I believe knex puts down too much work towards handling pooling. We have swapped pooling libs several times already. Why not rely on something that is already well tested?

At least this time swapping goes to library that is well tested and that we can easily debug and fix.

I don't believe that there are many well tested components in JS world (except for Tarn and Objection), generic pool and pool2 were supposed to be, but clearly they were not. Maybe it would be worth looking how pooling is done in sequelize if they are using driver builtin pools?

**Comment by elhigu** [comment]
Merged #2450

**Comment by LewisJEllis** [comment]
> Maybe it would be worth looking how pooling is done in sequelize if they are using driver builtin pools?

They don't, they do their own generic-pool-based connection pool on top of the drivers.
